### ISSUE TITLE
Transformation pass discards layout information

### ISSUE THREAD

**Issue by eendebakpt** [Qiskit/qiskit-terra#8622]
### Environment

- **Qiskit Terra version**:  0.21.2
- **Python version**:  3.10
- **Operating system**:  Windows

### What is happening?

Applying a transformation pass discards information about the layout of a quantum circuit.

### How can we reproduce the issue?

A minimal example is:
```
from qiskit.circuit.quantumcircuit import QuantumCircuit
from qiskit.compiler import transpile
from qiskit.transpiler import CouplingMap
from qiskit.transpiler.passmanager import PassManager
from qiskit.transpiler.passes import Decompose

qc=QuantumCircuit(3)
qc.cx(0, 2)    

coupling_map = CouplingMap([[0,1], [1,2],  [1,0], [2,1]])
print(qc.draw())    

qc_base = qc
qc_base = transpile(qc_base, basis_gates=['cz', 'cx', 'u3', 'u2', 'u1'], coupling_map=coupling_map,  optimization_level=2)
print(qc_base.draw()) 


d=Decompose(gates_to_decompose=None)
pm = PassManager([d, ])
qc_base2= pm.run(qc_base)
print(qc_base2.draw()) # here we seem to have lost the mapping information
```
Output:
```
q_0: ──■──
       │  
q_1: ──┼──
     ┌─┴─┐
q_2: ┤ X ├
     └───┘
         ┌───┐
q_2 -> 0 ┤ X ├
         └─┬─┘
q_0 -> 1 ──■──
              
q_1 -> 2 ─────
              
     ┌───┐
q_0: ┤ X ├
     └─┬─┘
q_1: ──■──
          
q_2: ─────
```

### What should happen?

I want to apply a series of passes (including passes that deal with the coupling map) to obtain a circuit that I can execute on the specified topology, but with an unchanged qubit ordering.

It is fine if the passes change the order of the qubits (which happens in the minimal example for circuit `qc_base` where the qubits 0, 1, 2 are re-ordered to 1, 2, 0), as long as I have the information to map this circuit back to a circuit with the original qubit ordering.

**Comment by jakelishman** [comment]
There's a few "magic" properties on `QuantumCircuit` that are written out from similarly magic properties in the property sets of running pass managers.  The layout is one of them.

It's not immediately clear to me if there are significant problematic implications around `RunningPassManager.run` propagating these magic properties from an input circuit to an output circuit (if they don't appear in the output property set).  I _suspect_ the layout is probably ok, since in the preset pass managers, the layout is set very early on in the process, and subsequent passes don't invalidate it.  The other magic properties are to do with scheduling, and those probably can't be safely restored, because almost every pass would invalidate them.

As an immediate workaround, you can just do `qc_base2._layout = qc_base._layout` to restore the layout.

**Comment by kdk** [comment]
Agree. It looks like `PassManager.run` will only write `circuit._layout` if a layout is generated in the `property_set`, so at least it should be possible to carry `_layout` forward when no layout passes are run, and raise/warn if there's both a existing `._layout` and the one trying to be set by the `PassManager`.

https://github.com/Qiskit/qiskit-terra/blob/fb2e0e5e9459b8754e350d6a121c29a5a0e25065/qiskit/transpiler/runningpassmanager.py#L105

**Comment by flowerthrower** [comment]
Thanks for keeping this issue open. We have hit the same boundary while preparing [MQT Predictor v3](https://github.com/munich-quantum-toolkit/predictor).

### Predictor use case

MQT Predictor is an RL-based compiler where compilation is intentionally incremental. Each action runs one or a small group of passes in a fresh `PassManager`; the next action may come from Qiskit, TKET, or BQSKit.

Once a layout has been applied, subsequent actions need to continue from the existing `TranspileLayout`, since it connects the physical circuit back to the original logical qubits.

With Qiskit 2.5.2, I reproduced two variants of the problem:

- Running a layout-preserving transformation such as `Decompose` in a new `PassManager` returns a circuit with `layout=None`.
- Running `SabreSwap` in a new `PassManager` produces a layout relative to the already-physical input circuit rather than composing it with the original layout.

Predictor currently works around this by carrying the layout separately, manually updating routing metadata, and assigning it back through `QuantumCircuit._layout`. This works, but duplicates Qiskit's layout bookkeeping and relies on private state.

### Existing design

I understand that this is not equivalent to carrying arbitrary `PassManager` analysis state.

The earlier discussion here already distinguishes layout from scheduling information. I also saw the more recent [discussion in #16100](https://github.com/Qiskit/qiskit/issues/16100#issuecomment-4342793612) explaining that entering a new `PassManager` has historically represented a complete new compilation, so changing the default behavior requires care.

Since this issue was opened, several useful building blocks have landed (there might be more?):

- #11399 added composition of an existing `final_layout` in routing passes; #14919 later corrected the composition order.
- #14825 added initial `PropertySet` handling at `PassManager` entry.
- #14826 added `TranspileLayout.from_property_set()` and `write_into_property_set()`.
- #14778 introduced the Rust `TranspileLayout`, and #14904 began using it in `ApplyLayout` as part of moving layout state into the IR.
- #12534 explored a broader representation change, but was closed as outdated.

### Proposal

The smallest solution I can see is an explicit opt-in for continuing from the input circuit's existing layout state. At a high level, the `PassManager` frontend could seed only the layout-related entries through the existing conversion method:

```python
# Illustrative only; option name is deliberately not prescribed.
layout_keys = ("layout", "final_layout", "virtual_permutation_layout")
has_explicit_layout = any(self.property_set.get(key) is not None for key in layout_keys)

if (
    continue_existing_layout
    and input_program.layout is not None
    and not has_explicit_layout
):
    input_program.layout.write_into_property_set(self.property_set)
```

The existing passes would continue to update the `PropertySet`, routing passes would use their existing composition behavior, and the existing `TranspileLayout.from_property_set()` backend path would construct the output layout.

### Implications

- The default fresh-compilation behavior would remain unchanged.
- Only layout state would be continued; scheduling and unrelated analyses would not be copied.
- No new permutation-composition implementation or pass classification would be needed.
- The initial scope could be circuits carrying a `TranspileLayout` produced after `ApplyLayout`. The opt-in would rely on the existing pass contract: passes that change qubit identities or introduce permutations must update the corresponding layout state.
- Starting another initial-layout stage from an already-laid-out circuit could remain unsupported or raise/warn, consistent with the earlier suggestion in this issue.
- An explicitly supplied layout-bearing `property_set` should take precedence. Otherwise, the complete layout bundle from `input_program.layout`, including the original-qubit metadata, should be seeded atomically.
- Custom passes that change qubits or introduce implicit permutations without updating the layout would retain their existing responsibility.

I'd be happy to contribute a fix. Let me know if this goes in the right direction or if you have any other ideas.

**Comment by jakelishman** [comment]
Patrick: in your particular use-case, semantically are you not just trying to pretend that the _whole_ pass-manager state is continuous, if you back out to another library's passes?  Something feels off about an option as specific as `continue_existing_layout`: would we need a new option in the future if somebody wants to break between two different passes that set state?

You're already reaching into the compilation internals here, and trying to thread through state.  The mechanism we have for that is that `PassManager.run` explicitly allows you to seed it with a `PropertySet` that it should continue from (and you can just take the `.property_set` property from the PM after execution finishes).  If you're occasionally getting the layout/routing from another library, then it should be valid to build the Qiskit `TranspileLayout`, call its `write_into_property_set` method passing the `PropertySet` you're holding between PMs, then seed that.  All that stuff is already public API, and gives you more control (you're free to mess with the property set - it's just a dictionary).

Basically, I'm saying the general case of "continue this compilation" is spelled by doing
```python
property_set = PropertySet()
pms = [PassManager(...), PassManager(...), ...]   # can be dynamic
circuit = QuantumCircuit(...)

for pm in pms:
    circuit = pm.run(circuit, property_set=property_set)
    property_set = pm.property_set  # these would be the same object anyway
```
This requires you to pass only single circuits to `run` - the `list[QuantumCircuit]` form doesn't expose the multiple states.  Imo that's already a pretty simple API, and it would work even if you insert passes from other libraries in the middle, provided you convert any of their analyses (layout, etc) into our forms and write them into the `property_set` as appropriate.

We may well have to make the input initialisation function check the incoming `PropertySet` to make sure it doesn't clobber any incoming keys, but I don't think we need the `continue_existing_layout` argument you suggested: it's either already in there and we don't clobber it, or it isn't, and we infer a need to write out the input bits.

Btw, probably `QuantumCircuit.layout` should just be publicly mutable, perhaps with validation on set, if that's one of your concerns.

**Comment by flowerthrower** [comment]
Hi @jakelishman—thanks for the reply! I agree that `continue_existing_layout` feels too specific; using `property_set` as the explicit continuation mechanism is cleaner. At external-compiler boundaries, Predictor can translate the resulting permutation into its tracked `TranspileLayout`, create a fresh `PropertySet`, and seed only that layout state with `write_into_property_set()`.

One issue:I think the frontend overwrites the supplied original-qubit mapping and input-qubit count with metadata from the already-physical circuit, while retaining the original layout. This mixes physical and logical qubit identities and can make methods such as `final_index_layout()` fail.

I can narrow the PR to preserve these values when supplied and infer them only when unset. This removes the new option while leaving fresh runs unchanged.

Making `QuantumCircuit.layout` publicly assignable would let Predictor perform its remaining cross-library reattachment without accessing `_layout`, but that  can also be discussed separately.
